### PR TITLE
Async input

### PR DESCRIPTION
**Pull request by bitnom**
async tasks were getting blocked by input(). We solve by spreading the async virus. All the async stuff needs a refactor generally but that's for later.

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by ekzhu** [comment]
Thanks for the PR, I am wondering if we can combine the effort based on the work here #1240 ? I think human input can be added as a middleware. In that case we should probably focus on reviewing and enhanching #1240 and build new ConversableAgent features from there.

File: autogen/agentchat/conversable_agent.py [hunk 1]
```diff
@@ -1355,7 +1356,7 @@ async def a_get_human_input(self, prompt: str) -> str:
         Returns:
             str: human input.
         """
-        reply = input(prompt)
+        reply = await ainput(prompt)
```
Comment (davorrunje) [inline autogen/agentchat/conversable_agent.py:1359]: you cannot `await` in a `sync` function

File: autogen/misc_utils.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
+import asyncio
+import sys
+
+
+async def ainput(prompt: str) -> str:
```
Comment (davorrunje) [inline autogen/misc_utils.py:5]: The convention in this repo is to use `a_*` for names of `async` functions.

**Review by davorrunje — CHANGES_REQUESTED** [review]
Can you please write tests for the new code added?

**Comment by davorrunje** [comment]
Other than those details mentioned above, this is a very useful feature 👏

File: autogen/misc_utils.py [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    """
+    Async alternative to input()
+    Args:
+        prompt (str): Prompt for user input.
+
+    Returns: The stdin str.
+
+    """
+    await asyncio.get_event_loop().run_in_executor(None, lambda s=prompt: sys.stdout.write(s + " "))
```
[the remark this answers is not in the document]
Comment (bitnom) [reply autogen/misc_utils.py:14]: I think it is blocking. It's just usually too fast to make much of a difference. The original implementation is came from was meant for general use-cases. It's probably way overkill for this input prompt though.

File: autogen/misc_utils.py (hunk 2 above)
Comment (bitnom) [reply autogen/misc_utils.py]: fixed

File: autogen/agentchat/conversable_agent.py (hunk 1 above)
Comment (bitnom) [reply autogen/agentchat/conversable_agent.py]: It's not a sync function. I had created the new async method `a_get_human_input`

File: autogen/agentchat/conversable_agent.py [hunk 4]
```diff
@@ -1565,6 +1566,20 @@ def generate_init_message(self, **context) -> Union[str, Dict]:
         """
         return context["message"]
 
+    async def a_generate_init_message(self, **context) -> Union[str, Dict]:
```
[the remark this answers is not in the document]
Comment (bitnom) [reply autogen/agentchat/conversable_agent.py]: Will it? That's the same hint the original sync method had. I just copied it over. Anyway, I updated them both.

[the remark this answers is not in the document]
Comment (davorrunje) [reply autogen/agentchat/conversable_agent.py]: It is easy to check and yes, it will complain.

File: autogen/agentchat/conversable_agent.py (hunk 1 above)
Comment (davorrunje) [reply autogen/agentchat/conversable_agent.py]: sorry, my mistake

File: autogen/misc_utils.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (bitnom) [reply autogen/misc_utils.py]: yeah it was unnecessary overhead. I got rid of it.

**Comment by bitnom** [comment]
> Can you please write tests for the new code added?

should be good now

File: test/agentchat/test_async_get_human_input.py [hunk 5]
```diff
@@ -19,26 +21,26 @@
...
+        llm_config={"seed": 41, "config_list": config_list, "temperature": 0},
     )
 
     user_proxy = autogen.UserProxyAgent(name="user", human_input_mode="ALWAYS", code_execution_config=False)
 
-    async def custom_a_get_human_input(prompt):
-        return "This is a test"
-
-    user_proxy.a_get_human_input = custom_a_get_human_input
+    user_proxy.a_get_human_input = AsyncMock(return_value="This is a test")
```
Comment (davorrunje) [inline test/agentchat/test_async_get_human_input.py:39]: The tests you need:
1. mock the response from `sys.stdin.readline` and test `a_input()`
2. mock `a_input` and test `a_get_human_input`
3. mock `a_get_human_input` and test `a_initiate_chat`

**Comment by gagb** [comment]
Closing because of inactivity

**Pull request closed without merging**